src/librustc/ty/mod.rs

```diff
@@ -136,6 +136,9 @@ pub struct ResolverOutputs {
     /// Extern prelude entries. The value is `true` if the entry was introduced
     /// via `extern crate` item and not `--extern` option or compiler built-in.
     pub extern_prelude: FxHashMap<Name, bool>,
+    /// All crates that ended up getting loaded.
+    /// Used to determine which `--extern` entries are unused
+    pub used_crates: Option<FxHashSet<Symbol>>,
```

Comment:
Putting this list into `ResolverOutputs` should be unnecessary, the lint can be reported before `ResolverOutputs` are constructed.
The whole lint logic should be able to live entirely inside `rustc_metadata`, I suggested the point to report it in https://github.com/rust-lang/rust/pull/69203#issuecomment-586691850 (`CrateLoader::postprocess`).

Reply:
@petrochenkov: Things are not actually done when `postprocess` is called - while we won't resolve any more crates, we may register additional names for crates that are already loaded.

Reply:
>we may register additional names for crates that are already loaded.

How?
(Ignoring rustdoc, which shouldn't report this lint.)

Reply:
If we find a previous crate when loading a crate: https://github.com/rust-lang/rust/blob/1572c433eed495d0ade41511ae106b180e02851d/src/librustc_metadata/creader.rs#L522

then we may map a different name in the extern prelude to an already loaded crate.

Reply:
`CrateLoader::load` can only happen before `postprocess`.
(Or rather it can happen later during AST lowering when resolving `std` paths, but if lowering actually loads new crates, it will break more than this lint.)
Could you give some examples of the false positives?